RMB-388: Add getById tx and lock variants

hjiebsco opened the pull request:
See https://issues.folio.org/browse/RMB-388

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2379,26 +2379,42 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
-        return;
+    Promise<PgConnection> promise = Promise.promise();
+    if (conn != null) {
+      if (conn.failed()) {
+        promise.fail(conn.cause());
+      } else {
+        promise.complete(conn.result().conn);
```
```suggestion
      promise.handle(conn.map(sqlConnection -> sqlConnection.conn));
```

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2379,26 +2379,42 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
       }
-      PgConnection connection = res.result();
+    } else {
+      getConnection(res -> {
+        if (res.failed()) {
+          promise.fail(res.cause());
+        } else {
+          promise.complete(res.result());
+        }
+      });
```
```suggestion
      getConnection(promise);
```

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2379,26 +2379,42 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
+        }
+      });
+    }
+    promise.future().onFailure(ex -> replyHandler.handle(Future.failedFuture(ex)))
```
```suggestion
    promise.future()
    .onFailure(ex -> replyHandler.handle(Future.failedFuture(ex)))
```

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2379,26 +2379,42 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
 
   /**
    * Get the jsonb by id.
+   * @param conn  if provided, the connection on which to execute the query on.
```
The description should explain all three cases:
* if provided, the connection to use
* if null, a new connection is created and automatically released afterwards
* if failed, the replyHandler is failed

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2379,26 +2379,42 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
+    .onSuccess(connection -> {
       String sql = SELECT + DEFAULT_JSONB_FIELD_NAME
           + FROM + schemaName + DOT + table
           + WHERE + ID_FIELD + "= $1";
+      if (lock) {
+        sql += " FOR UPDATE";
+      }
```
```suggestion
          + WHERE + ID_FIELD + "= $1"
          + (lock ? " FOR UPDATE" : "");
```
Creating the String in one go allows the compiler to implement it using a StringBuilder.

julianladisch replied:
Can you write a test with a real PostgreSQL (not a mock) to ensure that "FOR UPDATE" actually works:
On first connection run "SELECT ... FOR UPDATE".
On second connection run "SELECT" for that row, this will block until the first connection closes the transaction.
On first connection run "SELECT pg_sleep(0.1)", "UPDATE" the record, and COMMIT the transaction.
The second connection should return the new value from the UPDATE, not the old value.

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2429,7 +2449,34 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
+
+  /**
+   * Get the jsonb by id and return it as a String. Query will be executed on
+   * <code>conn</code> if provided.
```
The sentence `Query will be executed on <code>conn</code> if provided.` duplicates the information provided in the `@paran` entry and should be removed. The javadoc should be concise.

julianladisch commented on domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java:
```diff
@@ -2429,7 +2449,34 @@ private FacetManager buildFacetManager(CQLWrapper wrapper, QueryHelper queryHelp
+  }
+
+  /**
+   * Get the jsonb by id and return it as a String. Query will be executed on
+   * <code>conn</code> if provided. Selected row will be locked using
+   * <code>select ... for update<code> clause.
```
The key information should be in the beginning:

`Lock the row using <code>select ... for update</code> and return jsonb as String.`

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientTest.java:
```diff
@@ -709,5 +715,96 @@ public void getTotalRecordsTest() {
+    SQLConnection mockSQLConnection = new SQLConnection(mockPgConnection, null, null);
+    AsyncResult<SQLConnection> mockConn = Future.succeededFuture(mockSQLConnection);
+    // tests
+    pc.getByIdAsString(table, id, ar -> assertTrue(ar.succeeded()));
```
If the method calls the callback then `ar -> assertTrue(ar.succeeded())` will run.
If the method doesn't complete and doesn't call the callback then `ar -> assertTrue(ar.succeeded())` will NOT run but the unit test will succeed and does NOT report this incomplete execution.
Please use https://vertx.io/docs/vertx-unit/java/ or https://vertx.io/docs/vertx-junit5/java/ that provide test methods for such tests, for example testContext.asyncAssertSuccess() or vertxTestContext.succeeding().

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientTest.java:
```diff
@@ -709,5 +715,96 @@ public void getTotalRecordsTest() {
+  } 
+  
+  private void assertGetByIdAsString(AsyncResult<String> ar) {
+    assertTrue(ar.succeeded());
```
assertTrue(ar.succeeded())
should be handled by testContext.asyncAssertSuccess() or vertxTestContext.succeeding().
They also unwrap the result.

hjiebsco commented:
@julianladisch, thanks for the feedback. Please review it again. I moved the mock test to a new file. I also added an IT test case. I tweaked your proposed IT test logic a tiny bit because `select ... for update` prevents concurrent update, not select.

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java:
```diff
@@ -1483,6 +1483,57 @@ public void getByIdAsString(TestContext context) {
+              postgresClient.endTx(conn, done -> {
+                postgresClient.getByIdAsString(FOO, id, get1 -> {
+                  assertSuccess(context, get1);
+                  context.assertTrue(get1.result().contains("\"x1\""), get1.result());
```
This getByIdAsString should go before the endTx to avoid any race condition.

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java:
```diff
@@ -1483,6 +1483,57 @@ public void getByIdAsString(TestContext context) {
+                new CQLWrapper(criterion), true, rows -> {
+              assertSuccess(context, rows);
+              context.assertFalse(async2.isCompleted());
+              Row row = rows.result().iterator().next();
+              context.assertEquals(id, row.getUUID("id").toString());
```
This assertEquals is not needed, we can assume that postgresClient.update works correctly.

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java:
```diff
@@ -1483,6 +1483,57 @@ public void getByIdAsString(TestContext context) {
+            });
+          });
+          // before update
+          postgresClient.getByIdAsString(FOO, id, get0 -> {
+            assertSuccess(context, get0);
+            context.assertTrue(get0.result().contains("\"x0\""), get0.result());
```
This getByIdAsString with assert should be removed, we can assume that save works correctly.

julianladisch commented on domain-models-runtime/src/test/java/org/folio/rest/persist/PostgresClientIT.java:
```diff
@@ -1483,6 +1483,57 @@ public void getByIdAsString(TestContext context) {
+            });
+          });
+          // update first because it holds the lock
+          postgresClient.execute(conn, "SELECT pg_sleep(0.5)", delayRs -> {
```
I don't think that we need the pg_sleep. Instead we can simply run "SELECT 1" to make sure that the update x2 has reached the database and is blocked before we run update x1.

hjiebsco replied:
Can you clarify how to use `SELECT 1` to make sure that the update x2 has reached the db? We could query system tables to check locks (if that is what you mean) but it is probably overkill than `pg_sleep`.

julianladisch reviewed (approved):
Looks good, thanks!
